### ISSUE TITLE
Circular dependency between mso_schema and mso_schema_template results in duplicate and hidden resources

### ISSUE THREAD

**Issue by mike-guy** [CiscoDevNet/terraform-provider-mso#110]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

terraform version
Terraform v1.1.4
on windows_amd64
+ provider registry.terraform.io/ciscodevnet/mso v0.4.1

### MSO version

Nexus Dashboard
Version 2.0(2g)

Multi-Site Orchestrator
Version: 3.3(1e)

dCloud Cisco Nexus Dashboard Orchestrator for ACI Lab v1

### APIC version and APIC Platform for Site Level Resources

N/A

### Affected Resource(s)

* mso_schema 
* mso_schema_tempalte

### Expected Behavior
Expected that schema and templates can be declaratively declared in Terraform code without needing to perform workarounds.

### Actual Behavior
mso_schema creates additional hidden resources resulting in duplicate templates being created, one of which is not managed by Terraform state.

### Steps to Reproduce
Apply the following Terraform code.

```hcl
terraform {
  required_providers {
    mso = {
      source  = "CiscoDevNet/mso"
      version = "0.4.1"
    }
  }
}

provider "mso" {
  # Creds provided as environment variables
}

resource "mso_tenant" "tenant" {
  name         = "test_tenant"
  display_name = "test_tenant"
}

resource "mso_schema" "schema" {
  name          = "test_schema"
  tenant_id     = mso_tenant.tenant.id
  template_name = mso_schema_template.template.name
}

resource "mso_schema_template" "template" {
  name         = "test_template"
  display_name = "test_template"
  tenant_id    = mso_tenant.tenant.id
  schema_id    = mso_schema.schema.id
}
```

### Logic Issue
There is a circular dependency that causes a non-declarative approach to resource definition

* The resource mso_schema requires a template_name
* The resource mso_schema_template requires a schema_id

This would usually be provided by way of resource references. Whilst this appears to still work (as the values are known before apply), it causes a duplicate template to be created. The "mso_schema" automatically creates a template that is not managed by Terraform.

#### Applying the earlier template
```hcl
bug_testing> terraform apply

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # mso_schema.schema will be created
  + resource "mso_schema" "schema" {
      + id            = (known after apply)
      + name          = "test_schema"
      + template_name = "test_template"
      + tenant_id     = (known after apply)
    }

  # mso_schema_template.template will be created
  + resource "mso_schema_template" "template" {
      + display_name = "test_template"
      + id           = (known after apply)
      + name         = "test_template"
      + schema_id    = (known after apply)
      + tenant_id    = (known after apply)
    }

  # mso_tenant.tenant will be created
  + resource "mso_tenant" "tenant" {
      + description  = (known after apply)
      + display_name = "test_tenant"
      + id           = (known after apply)
      + name         = "test_tenant"

      + site_associations {
          + aws_access_key_id         = (known after apply)
          + aws_account_id            = (known after apply)
          + aws_secret_key            = (known after apply)
          + azure_access_type         = (known after apply)
          + azure_active_directory_id = (known after apply)
          + azure_application_id      = (known after apply)
          + azure_client_secret       = (known after apply)
          + azure_shared_account_id   = (known after apply)
          + azure_subscription_id     = (known after apply)
          + is_aws_account_trusted    = (known after apply)
          + security_domains          = (known after apply)
          + site_id                   = (known after apply)
          + vendor                    = (known after apply)
        }

      + user_associations {
          + user_id = (known after apply)
        }
    }

Plan: 3 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

mso_tenant.tenant: Creating...
mso_tenant.tenant: Creation complete after 1s [id=62164fef2e0000b90170b5d9]
mso_schema.schema: Creating...
mso_schema.schema: Creation complete after 0s [id=62164ff02e0000b90170b5da]
mso_schema_template.template: Creating...
mso_schema_template.template: Creation complete after 1s [id=test_template]

Apply complete! Resources: 3 added, 0 changed, 0 destroyed.
```

#### State file after apply

```hcl
bug_testing> terraform show
# mso_schema.schema:
resource "mso_schema" "schema" {
    id            = "62164ff02e0000b90170b5da"
    name          = "test_schema"
    template_name = "test_template"
    tenant_id     = "62164fef2e0000b90170b5d9"
}

# mso_schema_template.template:
resource "mso_schema_template" "template" {   
    display_name = "test_template"
    id           = "test_template"
    name         = "test_template"
    schema_id    = "62164ff02e0000b90170b5da" 
    tenant_id    = "62164fef2e0000b90170b5d9" 
}

# mso_tenant.tenant:
resource "mso_tenant" "tenant" {
    description  = jsonencode({})
    display_name = "test_tenant"
    id           = "62164fef2e0000b90170b5d9"
    name         = "test_tenant"

    user_associations {
        user_id = "48d105bdfbc49a5fcf39a10b961386e61dfe40085cb0315d818b761c5735dafa"
    }
}
```

#### Actual state of MSO

![image](https://user-images.githubusercontent.com/48997783/155349380-1f0de591-a747-471a-844a-a7a1cd2789e0.png)

**Comment by shrsr** [comment]
@mike-guy resource mso_schema creates a template in and of itself. If you only want to create one/first template use mso_schema. Resource mso_schema_template is used to create subsequent templates. Besides that, we don't create references between two resources such that they depend on each other.

Nevertheless, I tried using your configuration and received the following:
Error: Cycle: mso_schema.schema, mso_schema_template.template

Please let us know if you have any question.

**Comment by mike-guy** [comment]
Hi,

Yep that's exactly my point. The template that is created as part of "mso_schema" is essentially a "hidden" resource from a terraform perspective and cannot be configured via a terraform resource. This means any future items under that template end up having to use a data soirce rather than just "mso_schema.someschema.id". 

Mso_schema (in my opinion) should JUST create a schema. If a template is mandatory when the API call is made, this may present a challenge, but I'm sure it will be possible. 

I haven't poked around the code yet, but I suspect this is a workaround to do with MSO schemas requiring a template. 

I'm away for a bit but will have a closer look next week if I can.

**Comment by shrsr** [comment]
Hi Mike.

Based on the API of non-ND based MSO (versions prior to 3.3), it is required to add a template when creating a schema.
On NDO, we are evaluating whether the behavior is same. We'll come up with an alternative based on what we find.

**Comment by lhercot** [comment]
Hi Mike, we have decided to implement the ability to create more than 1 template using a block in the mso_schema resources to respond to this issue. We just merged a PR containing those changes (#135).
